src/hotspot/share/opto/divconstants.cpp

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "utilities/powerOfTwo.hpp"
+
+#include <limits>
+#include <type_traits>
+
+// Compute magic multiplier and shift constant for converting a 32/64 bit
+// division by constant into a multiply/shift series.
+//
+// (1) Theory:
+// Motivated by Henry S. Warren. 2012. Hacker's Delight (2nd. ed.). Addison-Wesley Professional.
+//
+// Given positive integers d <= N, call v the largest integer not larger than
+// N such that v + 1 is divisible by d.
+//
+// (a) For positive values c, m such that:
+//
+// m <= c * d < m + m / v
+//
+// We have:
+//
+// floor(x / d) = floor(x * c / m) for every integer x in [0, N]
+//
+// (b) For positive values c, m such that:
+//
+// m < c * d <= m + m / v
+//
+// We have:
+//
+// ceil(x / d) = floor(x * c / m) + 1 for every integer x in [-N, 0)
+//
```

Comment:
For the record, the domain for non-negative dividends can be extended to `[0, v + d)`, which is usually larger than `[0, N]`, since `v <= N < v + d`.
Similarly, the domain for negative dividends can be extended to `(-(v + d), 0)`.

Reply:
Indeed, however if we go from a given `N` then the solution would be the same.

Reply:
Nothing would change in the algorithm.
I just wanted to point it out to have a record here (and in the mailing list).

To the reviewers of the less math-oriented aspects: I'm pretty confident that this algorithm is correct, so I would approve it in isolation if this were possible ;-)